AP_Motors: Heli: move swash functionality down into swash lib

IamPete1 opened the pull request:
Moves a bunch of functionality down into the swash lib and tidies it up a bit. Should be functionally identical although I have not done significant testing yet. 

The only outstanding tidy up of the swash stuff is its use in `get_motor_mask` that is just wrong and should be removed. It should only return actual motors. I will do a PR for that at somepoint but I didn't want to confuse this PR with that issue.

bnsgeyer commented:
@IamPete1 this is great but pretty extensive. You will have to prove it is identical for all swash types for 3 and 4 servo swashplates. 

As for the motor mask, that is out of my wheelhouse. Why is it wrong for heli?  It always did seem strange that we used the motors and rc_speed that sets the update rate for multicopter motors.  But I assumed there was a good reason.  Before we do anything rash, I would like to understand the history behind this. I’m sure @tridge may be able to shed some light on it.

Gone4Dirt commented:
I am expanding on the ap motors heli example that pete fixed up:
#23656

That will test output equivilence for all swash plate types and all heli frame types and give us more confidence that this PR is equivilent.

I agree that that motor mask is wrong.  But as you say @IamPete1 we should keep that out of this PR, to keep this one as a pure move of functionaility with equivilant outputs.  Will make it easier to test.

IamPete1 commented:
rebased.

IamPete1 commented:
Rebased on top of https://github.com/ArduPilot/ardupilot/pull/24194 to check there is no change. With the default swash plate this results in identical PWM outputs. We need to add support for the other swashplate types to that test before we can confirm this is fully correct.

IamPete1 commented:
@bnsgeyer This should now be good to go. It adds testing of all swashplate types for both single and dual. Results are:
```
Single_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Single_Heli H1:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Single_Heli H3_140:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 1.000000
                Mot2: 1.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Single_Heli H3_120:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Single_Heli H4_90:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Single_Heli H4_45:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 1.000000
                Mot2: 1.000000
                Mot3: 1.000000
                Mot4: 0.000000
                Mot5: 1.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H1:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3_140:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3_120:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H4_90:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H4_45:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000
```

No change for the majority, we didn't quite make identical PWM outputs, we see 1us PWM changes on `Single_Heli H3_140`, `Single_Heli H4_45`, although the PWMs are integers the scaled values there calculated from do suffer from floating point accuracy. I suspect if we were to be able to compare them directly we would see very small changes throughout, because were comparing the integer PWM's we only see that difference when it happens to be close to the threshold between one integer and the next. 

This testing proves the output, setup and defualting is unchanged. Note that the test fails to setup the forth servo on the second swahplate on dual heli, the default channel is 8 which conflicts with RSC, this behavior is the same with and without this PR. The only thing not captured by the test is the mask value used in the `set_update_rate` call.

Gone4Dirt commented on libraries/AP_Motors/AP_MotorsHeli_Dual.h:
```diff
@@ -87,8 +87,8 @@ class AP_MotorsHeli_Dual : public AP_MotorsHeli {
-    AP_MotorsHeli_Swash        _swashplate1;        // swashplate1
-    AP_MotorsHeli_Swash        _swashplate2;        // swashplate2
+    AP_MotorsHeli_Swash _swashplate1 { CH_1, CH_2, CH_3, CH_7 }; // swashplate1
+    AP_MotorsHeli_Swash _swashplate2 { CH_4, CH_5, CH_6, CH_8 }; // swashplate2
```
Not a now thing, but I think a minor future improvement would be reorder the motor config in heli so the swashplate servo assignment of the first swash matches a single heli, just to make setup a little more consistent for the user.

Gone4Dirt commented on libraries/AP_Motors/AP_MotorsHeli_Swash.cpp:
```diff
@@ -86,130 +87,157 @@ const AP_Param::GroupInfo AP_MotorsHeli_Swash::var_info[] = {
+        _output[i] = 2.0f * _output[i] - 1.0f;
+
+        if (_make_servo_linear) {
+            _output[i] = get_linear_servo_output(_output[i]);
```
There is no change to get_linear_servo_output here so not an issue now, but we should remember that this is another variable to the outputs that we are not testing in the motors test tool.

Gone4Dirt reviewed (changes requested):
I have had a good look through and all looks pretty good to me.  Its a nice clean up.  There a few minor changes that I would appreciate if you could make.

IamPete1 commented on libraries/AP_Motors/AP_MotorsHeli_Swash.cpp:
```diff
@@ -86,130 +87,157 @@ const AP_Param::GroupInfo AP_MotorsHeli_Swash::var_info[] = {
+
+void AP_MotorsHeli_Swash::add_servo_angle(uint8_t num, float angle, float collective)
+{
+    add_servo_raw(num,
+                  cosf(radians(angle + 90)),
+                  cosf(radians(angle)),
+                  collective);
```
Its this way to match the one in motors matrix. 

https://github.com/ArduPilot/ardupilot/blob/277d16831d25cea9aac5e625dd173c78b8ae936c/libraries/AP_Motors/AP_MotorsMatrix.cpp#L540-L541

Gone4Dirt replied:
Then the one in motors matrix is convoluted IMHO and is harder to work back to the conventions used.

IamPete1 replied:
@Gone4Dirt has volunteered to update this and the one it motors in a future PR.

IamPete1 commented on libraries/AP_Motors/examples/AP_Motors_test/run_heli_comparison.py:
```diff
@@ -94,7 +95,31 @@ def get_fields(self):
 # ==============================================================================
 
-frame_class_lookup = {6:'Single_Heli', 11:'Dual_Heli', 13:'Heli_Quad'}
+frame_class_lookup = {6:'Single_Heli', 11:'Dual_Heli'}
```
Because it can't do swash, so it would need some special handling for that case, this PR doesn't change anything in heli-quad so it seemed the path of least resistance.

Gone4Dirt replied:
If you are going to change what was already there then I don't think just deleting the old parts of a test, that we are trying to standardise, is appropriate.  The special handling should be added to this PR.

bnsgeyer commented:
@IamPete1 i was looking over the different swash types. What is the H3?  Is that the swash where the user can set the servo positions?  What servo positions do you have set?  Since that is a user settable swash, I would suggest some non standard positions. Just to test the equations to ensure they are correct.

bnsgeyer commented:
@IamPete1 i would sugggest changing the phase as well since that is a feature of the generic 3 servo swash. In the future I want to add phase to all of the swash types as this allows users to remove control coupling of the rotor due to high stiffness rotors.

IamPete1 commented:
Rebased and fixed up the python issues.

@bnsgeyer Correct, H3 is the user configurable servo positions. The tests  uses the defaults of -60, 60 and 180. The new `add_servo_angle` function is based of the way H3 used to work, so the maths is the same.

H3 used to be:
```
        _rollFactor[CH_1] = cosf(radians(_servo1_pos + 90 - _phase_angle));
        _rollFactor[CH_2] = cosf(radians(_servo2_pos + 90 - _phase_angle));
        _rollFactor[CH_3] = cosf(radians(_servo3_pos + 90 - _phase_angle));
        _pitchFactor[CH_1] = cosf(radians(_servo1_pos - _phase_angle));
        _pitchFactor[CH_2] = cosf(radians(_servo2_pos - _phase_angle));
        _pitchFactor[CH_3] = cosf(radians(_servo3_pos - _phase_angle));
```
It is now:
```
            add_servo_angle(CH_1, _servo1_pos - _phase_angle, 1.0);
            add_servo_angle(CH_2, _servo2_pos - _phase_angle, 1.0);
            add_servo_angle(CH_3, _servo3_pos - _phase_angle, 1.0);
```
So the `_servox_pos - _phase_angle` is the same. The cos and 90 deg offset for roll is moved down into the `add_servo_angle` function which does:
```
                  cosf(radians(angle + 90)),
                  cosf(radians(angle)),
```
The `add_servo_angle` function is also used by `H3 120`, `H4 90` and `H4 45` which have all been tested, so we can be sure that it is correct. So we just need to check that were giving the correct thing to the `add_servo_angle` call, which I think is covered by the default case.

It is quite tricky to add automated testing of the user settable params because we don't have the parameter library in the example. If you have a combination you would like tested I can do it manually by changing the default values in the table.

bnsgeyer commented:
@IamPete1  ok. Well then that sounds like it has been tested thoroughly since it use the same function. I would suggest adding two more tests to look at phase for the generic case. Make one for +10 deg and -10 deg of phase. That way we are checking phase. 
thanks!

IamPete1 commented:
@bnsgeyer Phase of -10:
```
Single_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000
```
Phase of plus 10:
```
Single_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000


Dual_Heli H3:
        Inputs max change:
                Roll: 0.000000
                Pitch: 0.000000
                Yaw: 0.000000
                Thr: 0.000000
        Outputs max change:
                Mot1: 0.000000
                Mot2: 0.000000
                Mot3: 0.000000
                Mot4: 0.000000
                Mot5: 0.000000
                Mot6: 0.000000
                Mot7: 0.000000
                Mot8: 0.000000
        Limits max change:
                LimR: 0.000000
                LimP: 0.000000
                LimY: 0.000000
                LimThD: 0.000000
                LimThU: 0.000000
```

No change in output for those particular phase angles. I'm sure if we looked for long enough we could find a phase angle that got the single PWM change we saw in some of the other tests.

bnsgeyer commented:
@IamPete1 I'm sure we could cases that might have small errors but I just wanted the phase of +10 and -10 to part of the standard test for the library.  It is just a good sanity check for the phase feature.

bnsgeyer reviewed (approved):
LGTM.  The automated test output looks fine.

rmackay9 commented on libraries/AP_Motors/examples/AP_Motors_test/AP_Motors_test.cpp:
```diff
@@ -85,6 +85,42 @@ void setup()
+
+                // Check that init worked
+                if (!motors->initialised_ok()) {
+                    ::printf("ERROR: swash=%0.0f re-initialisation failed\n", value);
```
It's unlikely that anyone will see this..
